# Lab 02 - Bài tập

---

## THÔNG TIN

- Họ tên: Lăng Hoàng Sơn
- MSV: 23682371
- Ngày hoàn thành: 11/1/2025


---
## NUMPY

Bạn hãy phát sinh ngẫu nhiên mảng ndarray có tên `arr` gồm 100 phần tử số nguyên có giá trị từ 0..1000. Thực hiện các thao tác sau:

### 1. Hiển thị các phần tử thỏa yêu cầu dưới đây

- 10 phần tử đầu tiên
- 5 phần tử cuối cùng
- Các phần tử thỏa mãn điều kiện chia hết cho 5 và lớn hơn 100

In [ ]:
import numpy as np

mang = np.random.randint(0, 1001, size = 100)
print(mang)
print("\n")

print("Hien thi 10 phan tu dau tien")
print(mang[0:10])
print("\n")

print("Hien thi 5 phan tu cuoi cung")
print(mang[-5:])

print("\n")

print("Hien thi cac phan tu thoa man dieu kien")
# Cach 1
for i in mang:
  if(i%5 == 0 and i > 100):
    print(i)
#Cach 2
print(mang[(mang % 5 == 0) & (mang > 100)])

[442 685 361 684 716 626 955 102 769 850 273 721 144 160 577 952  64  45
 928 837 949 636 693 717 543 172 703 804 871 504 720 588 516 409 701 420
 900 180 619 755 692 595 449 216 921 883 282 461 148 312 564 398 406 983
 177 363 756  69 114 521 443 588 133 681 526 873 297 871 540 413 197 770
 973 578 835 530 443 996 425 231  24 484 701 237 333 795  99 620 823 139
 638  75  44 996 905 598 826 680 797 209]


Hien thi 10 phan tu dau tien
[442 685 361 684 716 626 955 102 769 850]


Hien thi 5 phan tu cuoi cung
[598 826 680 797 209]


Hien thi cac phan tu thoa man dieu kien
685
955
850
160
720
420
900
180
755
595
540
770
835
530
425
795
620
905
680
[685 955 850 160 720 420 900 180 755 595 540 770 835 530 425 795 620 905
 680]


### 2. Cho biết mảng trên có bao nhiêu phần tử có giá trị nhỏ hơn 100. Hiển thị các phần tử đó

In [ ]:
mang_nho_hon_100 = mang[mang < 100]
mang_nho_hon_100.sort()
print(mang_nho_hon_100)



[24 44 45 64 69 75 99]


### 3. Thực hiện thống kê:

- Tính tổng mảng
- Tìm giá trị trung bình
- Tìm giá trị trung vị
- Tìm Q1, Q2, Q3
- Tìm phân vị 20 (percentile 20)

In [ ]:
import numpy
tong_mang = sum(mang)
print(tong_mang)
trungbinh_mang = mang.mean()
print(trungbinh_mang)
trungvi_mang = np.median(mang)
print(trungvi_mang)
Q1 = np.percentile(mang, 25)
Q2 = np.percentile(mang, 50)
Q3 = np.percentile(mang, 75)
print(Q1,Q2,Q3)
Q20 = np.percentile(mang,20)
print(Q20)

50415
504.15
484.5
273.0 484.5 743.0
244.6


---

## PANDAS

Thông tin các bộ dữ liệu:

- `car-sales.csv`: bộ dữ liệu chứa thông tin của các ôtô cũ ở một cửa hàng


### 1. Đọc dữ liệu

In [ ]:
# YOUR CODE HERE
import pandas as pd
grades = pd.read_csv('car-sales.csv', header = 0)

### 2. Dữ liệu có bao nhiêu dòng, bao nhiêu cột

Đáp án: (10 dòng, 5 cột)

In [ ]:
# YOUR CODE HERE
grades.shape
grades.head(11)

Manufacturer Colour  Odometer  Doors    Price
0       Toyota  White    150043      4   4000.0
1        Honda    Red     87899      4   5000.0
2       Toyota   Blue     32549      3   7000.0
3          BMW  Black     11179      5  22000.0
4       Nissan  White    213095      4   3500.0
5       Toyota  Green     99213      4   4500.0
6        Honda   Blue     45698      4   7500.0
7        Honda   Blue     54738      4   7000.0
8       Toyota  White     60000      4   6250.0
9       Nissan  White     31600      4   9700.0

- Định tính (phân loại): Manufacturer, Colour
- Định lượng: Odemeter, Price, Doors

In [ ]:
# YOUR CODE HERE
# grades.head(11)
# data_types = grades.dtypes

# print(data_types)
import pandas as pd

data_types = grades.dtypes


print(data_types)

dinhtinh = data_types[data_types == 'object'].index.tolist()
dinhluong = data_types[data_types != 'object'].index.tolist()

print("\nĐịnh tính :", dinhtinh)
print("Định lượng :", dinhluong)

Manufacturer     object
Colour           object
Odometer          int64
Doors             int64
Price           float64
dtype: object

Định tính : ['Manufacturer', 'Colour']
Định lượng : ['Odometer', 'Doors', 'Price']


### 3. Trong dữ liệu có thông tin của bao nhiêu xe?

Đáp án: 10 xe

In [ ]:
# YOUR CODE HERE
dem_soluongxe = len(grades['Manufacturer'])
print(dem_soluongxe)
# YOUR CODE HERE
dem_soluongxe = grades['Manufacturer'].count()
print(dem_soluongxe)


10
10


### 4. Đếm số xe theo theo hãng

Đáp án: Toyota: 4 xe, Honda: 3 xe, Nissan: 2 xe, BMW: 1 xe

In [ ]:
# YOUR CODE HERE
dem_soluongxe = grades['Manufacturer'].value_counts()
print(dem_soluongxe)

Manufacturer
Toyota    4
Honda     3
Nissan    2
BMW       1
Name: count, dtype: int64


### 5. Đếm số xe theo màu


In [ ]:
# YOUR CODE HERE
dem_soluongxe = grades['Colour'].value_counts()
print(dem_soluongxe)

Colour
White    4
Blue     3
Red      1
Black    1
Green    1
Name: count, dtype: int64


### 6. Tổng giá trị các xe trong cửa hàng

Đáp án: 76450 USD


In [ ]:
# YOUR CODE HERE
tongtien = grades['Price'].sum()
print(tongtien,"USD")

76450.0 USD


### 7. Top 3 xe có giá cao nhất


In [ ]:
# YOUR CODE HERE
top3 = grades.nlargest(3, 'Price')
print(top3)

  Manufacturer Colour  Odometer  Doors    Price
3          BMW  Black     11179      5  22000.0
9       Nissan  White     31600      4   9700.0
6        Honda   Blue     45698      4   7500.0


### 8. Top 3 xe đã chạy nhiều cây số (km)nhất (Odometer)

Đáp án:
- Xe Nissan, White, 213095 km
- Xe Toyota, White, 150043 km
- Xe Toyota, Green, 99213 km

In [ ]:
# YOUR CODE HERE
grades.head(11)
top3 = grades.nlargest(3, 'Odometer')
print(top3[['Manufacturer','Colour','Odometer']])

  Manufacturer Colour  Odometer
4       Nissan  White    213095
0       Toyota  White    150043
5       Toyota  Green     99213


### 9. Khảo sát xe của hãng Honda

- Tính tỷ lệ xe Honda trong dữ liệu của cửa hàng (0.3)
- Cho biết giá trung bình của xe hãng Honda của cửa hàng (6500)
- Cho biết các xe của hãng Honda trong cửa hàng có những màu sắc nào (Red, Blue)

In [ ]:
# YOUR CODE HERE
grades.head(11)
tile_Honda = (grades['Manufacturer'] == 'Honda').mean()
print(tile_Honda)
tile_Honda = (grades['Manufacturer'] == 'Honda').mean()
print(tile_Honda)

honda = grades[grades['Manufacturer'] == 'Honda']
gia_trungbinh_Honda = honda['Price'].mean()
print(gia_trungbinh_Honda)
#Cach 2
# mau_sac_Honda = honda['Colour'].unique()
# print(mau_sac_Honda)

# tile_Honda = (grades['Manufacturer'] == 'Honda').mean()
# print("Tỷ lệ xe Honda:", tile_Honda)

# gia_trungbinh_Honda = grades[grades['Manufacturer'] == 'Honda']['Price'].mean()
# print("Giá trung bình xe Honda:", gia_trungbinh_Honda)

# mau_sac_Honda = grades.loc[grades['Manufacturer'] == 'Honda', 'Colour'].unique()
# print("Màu sắc xe Honda:", mau_sac_Honda)

Tỷ lệ xe Honda: 0.3
Giá trung bình xe Honda: 6500.0
Màu sắc xe Honda: ['Red' 'Blue']


### 10. Tính toán các giá trị sau của  thuộc tính giá ('Price')

- Giá trung bình các xe (7645)
- Giá trị trung vị các xe (6625)
- Miền giá trị của xe (18500)
- Miền phân vị của xe (2750)
- Điểm phân vị 0.4 và 0.6 của giá ([5750, 7000])

In [ ]:
# YOUR CODE HERE
grades.head(11)
giatri_trungbinh = grades['Price'].mean()
print("Giá trung bình các xe ",giatri_trungbinh)
giatri_trungvi = grades['Price'].median()
print("Giá trị trung vị các xe", giatri_trungvi)

#Miền giá trị là sự chênh lệch giữa giá trị lớn nhất và nhỏ nhất trong dữ liệu đó.
giatri_min = grades['Price'].min()
giatri_max = grades['Price'].max()
giatri_mien = giatri_max - giatri_min
print("Miền giá trị của xe",giatri_mien)
#Miền phân vị được tính bằng sự chênh lệch giữa phân vị 3 (Q3) và phân vị 1 (Q1):
giatriphanvi_min = np.percentile(grades.Price, 25)
giatriphanvi_max = np.percentile(grades.Price, 75)
giatriphanvi_mien = giatriphanvi_max - giatriphanvi_min
print("Miền phân vị của xe :",giatriphanvi_mien)
diemphanvi_04 = np.percentile(grades.Price, 40)
diemphanvi_06 = np.percentile(grades.Price, 60)
print("Điểm phân vị 0.4 và 0.6 của giá :",diemphanvi_04,diemphanvi_06 )


Giá trung bình các xe  7645.0
Giá trị trung vị các xe 6625.0
Miền giá trị của xe 18500.0
Miền phân vị của xe : 2750.0
Điểm phân vị 0.4 và 0.6 của giá : 5750.0 7000.0


---